Fix audio detection for IP4m-1041 Amcrest camera

flacjacket opened the pull request:
## Proposed change

The IP4m-1041 uses the `AudioIntensity` event code to signal an audio detection. The current audio detection event looks for the `AudioMutation` event code. Modify the event code handling to be able to check for multiple event codes and setup the audio detection event to monitor both codes.

## Type of change

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #77350
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist

- [ ] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

flacjacket committed:
Fix audio detection for IP4m-1041 Amcrest camera

flacjacket commented on homeassistant/components/amcrest/binary_sensor.py:
```diff
@@ -211,13 +211,18 @@ async def _async_update_others(self) -> None:
         try:
-            self._attr_is_on = (
-                len(await self._api.async_event_channels_happened(event_code)) > 0
+            self._attr_is_on = any(
```
I tried first doing this with a generator rather than a list comprehension, but then mypy complains with:
```
homeassistant/components/amcrest/binary_sensor.py:221: error: Argument 1 to "any" has incompatible type "AsyncGenerator[bool, None]"; expected "Iterable[object]"  [arg-type]
```
Let me try the generator instead, and silence mypy if needed...

flacjacket committed:
Use generator rather than list comprehension

flacjacket committed:
Fix mypy

Without the `type: ignore` in `binary_sensor.py`, mypy complains:
```
error: Argument 1 to "any" has incompatible type "AsyncGenerator[bool, None]"; expected "Iterable[object]"  [arg-type]
```
But mypy is ok if the `any` call takes a list comprehension with the
same arguments.  That however sets off pylint, so instead use the
generator with mypy silenced.

MartinHjelmare commented on homeassistant/components/amcrest/binary_sensor.py:
```diff
@@ -211,13 +211,16 @@ async def _async_update_others(self) -> None:
+        if (event_codes := self.entity_description.event_codes) is None or len(
+            event_codes
+        ) == 0:
             _LOGGER.error("Binary sensor %s event code not set", self.name)
```
Side note: This would be a developer error and then we should raise an exception, not log an error.

flacjacket replied:
That sounds good, I changed this over to raise a ValueError.

MartinHjelmare commented on homeassistant/components/amcrest/binary_sensor.py:
```diff
@@ -211,13 +211,16 @@ async def _async_update_others(self) -> None:
             return
 
-        if (event_code := self.entity_description.event_code) is None:
+        if (event_codes := self.entity_description.event_codes) is None or len(
```
`not event_codes` is simpler than checking length of the container. All Python containers support truthy check by default to mean empty / not empty.

flacjacket replied:
Sounds good, changed.

MartinHjelmare commented on homeassistant/components/amcrest/binary_sensor.py:
```diff
@@ -211,13 +211,16 @@ async def _async_update_others(self) -> None:
 
         try:
-            self._attr_is_on = (
+            self._attr_is_on = any(  # type: ignore[arg-type]
```
Why do we need to ignore type here?

flacjacket replied:
I think there is some problem with mypy being able to reason through the types with the async expression inside the generator. Without this, I get:
```
error: Argument 1 to "any" has incompatible type "AsyncGenerator[bool, None]"; expected "Iterable[object]"  [arg-type]
```
I can change the generator to a list comprehension and then mypy is happy without the `type: ignore` here, however, that upsets pylint. I figured it would be better to use the generator and silence mypy than to use the list comprehension and silence pylint.

MartinHjelmare replied:
Ok.

MartinHjelmare commented on homeassistant/components/amcrest/binary_sensor.py:
```diff
@@ -39,7 +39,7 @@
     """Describe Amcrest sensor entity."""
 
-    event_code: str | None = None
+    event_codes: list[str] | None = None
```
Should it be a set instead? The codes should be unique, right?

flacjacket replied:
Good idea, changed over to a set.

flacjacket committed:
Address review comments

Change list to set, truthy-ness check on maybe container, and logging to
exception

MartinHjelmare reviewed (approved):
Thanks!
